File: step_01_basics/1.1_logic_building/P004_pattern_4_right_angled_number_triangle_ii/solution.py
"""
Problem: Pattern 4 - Right Angled Number Triangle II
Difficulty: EASY | XP: 10

Print a right-angled triangle where row i prints the number i repeated i times.
Example (N=4):
1
2 2
3 3 3
4 4 4 4
"""


# ============================================================
# APPROACH 1: BRUTE FORCE -- Explicit Nested Loops
# Time: O(N^2)  |  Space: O(1)
# Outer loop for rows, inner loop prints i exactly i times.
# ============================================================
def brute_force(n: int) -> None:
    for i in range(1, n + 1):
        for j in range(1, i + 1):
            if j > 1:
                print(" ", end="")
            print(i, end="")
        print()


# ============================================================
# APPROACH 2: OPTIMAL -- String Repeat per Row
# Time: O(N^2)  |  Space: O(N) for current row string
# Use string multiplication to build each row in one expression.
# ============================================================
def optimal(n: int) -> None:
    for i in range(1, n + 1):
        print(" ".join([str(i)] * i))


# ============================================================
# APPROACH 3: BEST -- Full Grid String, Single Print
# Time: O(N^2)  |  Space: O(N^2)
# Build entire triangle as one string, single I/O call.
# ============================================================
def best(n: int) -> None:
    rows = [" ".join([str(i)] * i) for i in range(1, n + 1)]
    print("\n".join(rows))


if __name__ == "__main__":
    test_cases = [1, 3, 5]
    print("=== Pattern 4 - Right Angled Number Triangle II ===")

    for n in test_cases:
        print(f"\nN = {n}")
        print("--- Brute Force ---")
        brute_force(n)
        print("--- Optimal ---")
        optimal(n)
        print("--- Best ---")
        best(n)

File: step_01_basics/1.1_logic_building/P004_pattern_4_right_angled_number_triangle_ii/Solution.java
/**
 * Problem: Pattern 4 - Right Angled Number Triangle II
 * Difficulty: EASY | XP: 10
 *
 * Print a right-angled triangle with sequential numbers 1, 2, 3, ... across rows.
 *
 * @author DSA_Nova
 */
public class Solution {

    // ============================================================
    // APPROACH 1: NESTED LOOPS WITH RUNNING COUNTER
    // Time: O(N^2)  |  Space: O(1)
    // Maintain a counter that increments on every print.
    // ============================================================
    static class RunningCounter {
        public static void printPattern(int n) {
            int num = 1;
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= i; j++) {
                    System.out.print(num);
                    if (j < i) System.out.print(" ");
                    num++;
                }
                System.out.println();
            }
        }
    }

    // ============================================================
    // APPROACH 2: FORMULA-BASED (Compute Start per Row)
    // Time: O(N^2)  |  Space: O(1)
    // Row i starts at i*(i-1)/2 + 1. No running counter needed.
    // ============================================================
    static class FormulaBased {
        public static void printPattern(int n) {
            for (int i = 1; i <= n; i++) {
                int start = i * (i - 1) / 2 + 1;
                for (int j = 0; j < i; j++) {
                    if (j > 0) System.out.print(" ");
                    System.out.print(start + j);
                }
                System.out.println();
            }
        }
    }

    public static void main(String[] args) {
        int[] tests = {1, 3, 5};

        for (int n : tests) {
            System.out.println("=== Pattern 4: N = " + n + " ===");

            System.out.println("--- Running Counter ---");
            RunningCounter.printPattern(n);

            System.out.println("--- Formula Based ---");
            FormulaBased.printPattern(n);

            System.out.println();
        }
    }
}
